neuralmonkey/attention/scaled_dot_product.py

```diff
@@ -23,6 +24,143 @@
 # pylint: enable=invalid-name
 
 
+def split_for_heads(x: tf.Tensor, n_heads: int, head_dim: int) -> tf.Tensor:
+    """Split a tensor for multi-head attention.
+
+    Split last dimension of 3D vector of shape (batch, time, dim) and return
+    a 4D vector with shape (batch, n_heads, time, dim/n_heads)
+    """
+    x_shape = tf.shape(x)
+    x_4d = tf.reshape(tf.expand_dims(x, 2),
+                      [x_shape[0], x_shape[1], n_heads, head_dim])
+
+    return tf.transpose(x_4d, perm=[0, 2, 1, 3])
+
+
+def mask_weights(weights_4d: tf.Tensor, mask: tf.Tensor) -> tf.Tensor:
+    """Apply mask to softmax weights and renormalize."""
```

Comment:
docstring (args weights_4d: softmax weights to renormalize + přesunout shape z komentáře, mask: float Tensor of zeros and ones, specifies valid positions in the weight tensor + přesunout shape z komentáře nííže.
returns: Renormalized distribution over valid positions, has the same shape as weights_4d